### PR TITLE
backup: watch store state and retry backup immediately

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/52534
close https://github.com/pingcap/tidb/issues/53519
Problem Summary:

### What changed and how does it work?
1. reuse `storewatch` to observe store change event to retry backup.
2. refine main backup code for easy to test.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Leavrth — APPROVED** [review]

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -880,7 +1122,9 @@ func (bc *Client) OnBackupResponse(
 	resp := r.GetResponse()
 	storeID := r.GetStoreID()
 	if resp.GetError() == nil {
+		start := time.Now()
 		pr, err := globalProgressTree.FindContained(resp.StartKey, resp.EndKey)
+		logutil.CL(ctx).Debug("find the range tree contains reponse ranges", zap.Duration("take", time.Since(start)))
```
[the remark this answers is not in the document]
Comment (3pointer) [reply br/pkg/backup/client.go]: I think it's okay to keep the debug log here

File: br/pkg/backup/client.go [hunk 2]
```diff
@@ -803,39 +1057,27 @@ func (bc *Client) BackupRanges(
...
-	// }()
-
 	globalProgressTree, err := bc.BuildProgressRangeTree(ranges)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	err = bc.startMainBackupLoop(ctx, &globalProgressTree, replicaReadLabel, request, stateChan, progressCallBack)
+	stateNotifier := make(chan BackupRetryPolicy)
+	ObserveStoreChangesAsync(ctx, stateNotifier, bc.mgr.GetPDClient())
```
[the remark this answers is not in the document]
Comment (3pointer) [reply br/pkg/backup/client.go:1066]: If it gets stuck which means the receiver has stopped, due to current implement this must be backup main loop block at somewhere. so I think it's make sense to stuck if main loop blocked. what your concern here?

**Review by Leavrth — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**